teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+  such abstraction. Goal of this TEP is to let such authors notify about them
+  through CDEvents.
+- Give users the ability to select through configuration which event type they
+  would like to send, if any
+- Give Tekton operators the ability to remove event support from their
+  Tekton deployment
+
+### Non-Goals
+
+- Additional messaging features that may be implemented in future, like multiple
+  sinks, different event bindings, event signing and more
+- Changes to the operator to support disabling the events deployment are welcome,
+  but not part of this TEP
+
+### Use Cases
+
+- Notifications: Events produced by Tekton can be used to send notifications via
+  different channels (GitHub, Slack, etc)
+- Interoperability: CDEvents produced by Tekton can be consumed by other tools to
+  trigger the next step in the workflow. Similar events may be triggered by other
+  orchestrators like Jenkins and Argo Workflows, letting user switch between tools or
+  use a collection of them more easily
+- Observability and Audit Trail: Events generated by Tekton can be collected for
+  observability purposes along with events generated by other tools in the toolchain.
+  These stored events constitute an evidence store can be used to build audit trails,
+  visualize workflows, produce metrics
+- As a Task author I want to declare the purpose of my task according to the CDEvents
+  specification. As a Pipeline observer I want to know what tasks are performed in
+  CI/CD pipelines orchestrated by Tekton, according to the CDEvents specification.
+  Tekton is a non-opinionated pipeline orchestrator that allow workflow author to
+  automated any kind of Task in the CI/CD space (and beyond). Today the purpose of
+  the Tasks is usually transparent for Tekton: there is no concept of build, test,
+  CVE scan or deployment in Tekton. The Tekton community is working on the "artifact"
+  abstraction, but it's not available yet.
+
+### Requirements
+
+- Tekton operators can select which formats of events are produced by Tekton, with
+  one or more options available across Tekton v1, Tekton v2 and CDEvents
+- Tekton operators can disable events completely or enable events with separation
+  of permissions between the pipeline reconcilers and event emitters
+- Tekton task authors can send CDEvents using data available during the task execution
+- Tekton task authors can send CDEvents by exposing the required type and data through
+  an interface to be defined
+
+## Proposal
+
+The events controller is made of three controllers bundled together in one binary,
+one for each resource type: `TaskRun`, `PipelineRun` and `CustomRun`. It's deployed
+through a dedicated deployment and service, like today and it's part of the standard
+release. The operator may decide to provide a configuration option to skip its
+deployment.
+
+Each controller is read-only and uses and in-memory LRU cache to keep track of events
```

Comment:
how does HA work with this in memory cache?

Reply:
The HA relies on the shading of resources, which is based on the resource key `namespace/name`, which means that all reconciles for a certain resource are going to be served by the same controller, and thus find relevant data in the cache. 

A controller restart drops the cache, which means that some events may be sent twice (those for running resources).